// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

float xfail_rnz(float a, float b) {
  return __builtin_kvx_faddw(a, b, ".rnz"); // expected-error {{This builtin accepts a modifier string composed by: ['', '.', '.rn', '.rd', '.ru', '.rz'], ['', '.', '.s']}} expected-error {{cannot compile this builtin function yet}}
}
