// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

long long xfail_addcd(long long a, long long b) {
  return __builtin_kvx_addcd(a, b, ".ii"); // expected-error {{invalid carry_t modifier, expected "" or ".i"}} expected-error {{cannot compile this builtin function yet}}
}
