// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

float xfail_rna(float a, float b) {
  return __builtin_kvx_faddw(a, b, ".rna"); // expected-error {{This builtin accepts a modifier string composed by: ['', '.rz', '.', '.rn', '.rd', '.ru'], ['', '.', '.s']}} expected-error {{cannot compile this builtin function yet}}
}
