// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

typedef struct {long a, b;} s128;
typedef long __attribute__((__vector_size__(8 * sizeof(long)))) v8i64;

int wrong_ready(s128 *s, v8i64 *v){
  __builtin_kvx_ready(*s); // expected-error {{Aggregate types not supported in __builtin_kvx_ready}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_ready(*v); // expected-error {{Scalar size > 256 bits not supported in __builtin_kvx_ready}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_ready(); // expected-error {{__builtin_kvx_ready must have one to four arguments}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_ready(0, 1, 2, 3, 4); // expected-error {{__builtin_kvx_ready must have one to four arguments}} expected-error {{cannot compile this builtin function yet}}
  return __builtin_kvx_ready(0, 1, 2, 3); // expected-error {{Constant values should not be used in __builtin_kvx_ready}} expected-error {{cannot compile this builtin function yet}}
}
