// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

#include "vector-types.h"

v16i64 *toto;

void xfail_load1024(v4i64 *addr) {
  __kvx_x1024 tca = *((__kvx_x1024 *) toto);
  __builtin_kvx_xload1024q2(tca, addr, ".s"); // expected-error {{Expected to start with an uncached speculate modifier: '.u' or '.us'}} expected-error {{cannot compile this builtin function yet}}
  *((__kvx_x1024 *) toto) = tca;
}
