// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

#include "vector-types.h"

v8i16 xfail_addho(v8i16 a, v8i16 b) {
  return __builtin_kvx_addho(a, b, ".s.us"); // expected-error {{invalid saturate_t modifier, expected "", ".s" or ".us"}} expected-error {{cannot compile this builtin function yet}}
}
