// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

typedef struct {long a, b;} s128;
typedef long __attribute__((__vector_size__(8 * sizeof(long)))) v8i64;

void wrong_ready(s128 *s, v8i64 v, int sv, void * addr, int ready){
  __builtin_kvx_storew(sv, addr, 0, 1, 2); // expected-error {{No more than 4 arguments are allowed}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storew(sv, addr, ready, ready); // expected-error {{volatile flag should be a constant argument (e.g. 0 or 1)}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storew(sv, addr, *s); // expected-error {{ready argument should be a scalar}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storew(sv, addr, v); // expected-error {{ready argument size > 256 bits not supported}} expected-error {{cannot compile this builtin function yet}}
}
