// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -target-cpu kv3-2 -S -verify %s

typedef long __attribute__((__vector_size__(2 * sizeof(long)))) v2i64;
typedef long __attribute__((__vector_size__(4 * sizeof(long)))) v4i64;

void wrong_ready(int sv, void * addr, unsigned long cond, v4i64 big_sv, v2i64 mid_sv){
  __builtin_kvx_storecw(sv, addr, cond, ".mt"); // expected-error {{lsomask modifier is not available for this builtin. Try using __builtin_kvx_storec[64|128|256] instead}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storec256(big_sv, addr, cond, ".toto"); // expected-error {{Expects either a lsucond: ['.wlez', '.wnez', '.weqz', '.dgez', '.wgtz', '.even', '.dlez', '.dnez', '.wltz', '.deqz', '.odd', '.dgtz', '.wgez', '.dltz'] or a lsomask: ['.mt', '.mf']}} expected-error {{cannot compile this builtin function yet}}
}
