// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple kvx-kalray-cos -S -verify %s

typedef struct {long a, b;} s128;
typedef long __attribute__((__vector_size__(8 * sizeof(long)))) v8i64;

void wrong_ready(s128 *s, v8i64 v, int sv, void * addr, int ready, unsigned long cond){
  __builtin_kvx_storecw(sv, addr, cond, ".wnez", 0, 1, 2); // expected-error {{No more than 6 arguments are allowed}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storecw(sv, addr, cond, ".wnez", ready, ready); // expected-error {{volatile flag should be a constant argument (e.g. 0 or 1)}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storecw(sv, addr, cond, ".wnez", *s); // expected-error {{ready argument should be a scalar}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storecw(sv, addr, cond, ".wnez", v); // expected-error {{ready argument size > 256 bits not supported}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storecw(sv, addr, cond, ".mt"); // expected-error {{lsomask modifier is not available for this builtin. Try using __builtin_kvx_storec[64|128|256] instead}} expected-error {{cannot compile this builtin function yet}}
  __builtin_kvx_storecw(sv, addr, cond, ".toto"); // expected-error {{Expects a lsucond, one of: '.wlez', '.wnez', '.weqz', '.dgez', '.wgtz', '.even', '.dlez', '.dnez', '.wltz', '.deqz', '.odd', '.dgtz', '.wgez', '.dltz'}} expected-error {{cannot compile this builtin function yet}}
}
